### DISCUSSION TITLE
remove  one image from pdf

### DISCUSSION THREAD

**Discussion by pycgo** [pymupdf/PyMuPDF#1299] [Q&A] [answered]
i want to remove  one image from pdf  the code like this
```
doc=fitz.open(input_file)
images = doc.getPageImageList(0)
print(images)
```

the out is:
`[(7, 0, 2690, 4086, 1, 'DeviceGray', '', 'FXX1', 'CCITTFaxDecode'), (8, 0, 408, 697, 8, 'DeviceRGB', '', 'ntkoimage20210206232933', 'FlateDecode')]`

i want ntkoimage20210206232933   delete, how can i do

**Comment by JorjMcKie** [comment, marked as answer]
Use a redaction annotation:
1. determine bbox (rectangle) covered by image on page
2. create a redact annotation with the image's bbox
3. apply redaction

```python
bbox = page.get_image_rects(7)[0]  # get bbox
bbox = bbox.irect  # enlarge to include integer pixels
page.add_redact_annot(bbox)  # define redact annot
page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_REMOVE)
```

**Reply by pycgo** [reply]
Both pictures are deleted by your method, not what I want

**Reply by JorjMcKie** [reply]
If the images overlap, then this effect is inevitable. Do they?
There are ways to handle this, too.
Don't let me guess.

**Reply by pycgo** [reply]
my pdf file like this: https://github.com/pycgo/common-scripts/blob/master/python/testfile/111.pdf

 overlap? yes

**Reply by JorjMcKie** [reply]
You did not read the documentation on redaction annotations.
From there it is clear that **_everything_** (images or other stuff) overlapping a redact annot rect is being removed.
Snippet removing an image with recovering one that should be kept:
```python
>>> doc=fitz.open("test.pdf")
>>> page=doc[0]
>>> page.get_images()
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode'),
(7, 0, 1280, 720, 8, 'DeviceRGB', '', 'fzImg1', 'DCTDecode')]
>>> bbox1 = page.get_image_rects(5)[0]
>>> bbox2 = page.get_image_rects(7)[0]
>>> a = page.add_redact_annot(bbox2.irect)
>>> page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_REMOVE)
True
>>> page.insert_image(bbox1, xref=5)
5
>>> doc.ez_save("test2.pdf")
>>> 
```

**Reply by pycgo** [reply]
mycode
```
input_file = r'D:\code\pycode\pdf\output\111.pdf'
doc=fitz.open(input_file)
page = doc[0]
print(page.get_images())
bbox1 = page.get_image_rects(5)[0]
bbox2 = page.get_image_rects(359)[0]
a = page.add_redact_annot(bbox2.irect)
page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_REMOVE)
img1 = doc.extract_image(5)
page.insert_image(bbox1, xref=5, stream=img1["image"])
doc.save(r'output\222.pdf')
doc.close()
```
the error is
```
[(5, 0, 1345, 2036, 8, 'DeviceRGB', '', 'FXX1', 'DCTDecode'), (359, 0, 408, 697, 8, 'DeviceRGB', '', 'ntkoimage20210206232933', 'FlateDecode')]
Traceback (most recent call last):
  File "D:/code/pycode/pdf/dpdf.py", line 19, in <module>
    bbox2 = page.get_image_rects(359)[0]
IndexError: list index out of range
```

**Reply by JorjMcKie** [reply]
the reason is: image not shown on page

**Reply by pycgo** [reply]
Thank you, I know how to do it, delete the entire page picture, and then insert what I want to keep, it’s fine

**Reply by nissansz** [reply]
How to delete the entire page and then filter what you want to write back?
Any sample code?
